### ISSUE TITLE
Any plans to implement it in Rust?

### ISSUE THREAD

**Issue by Numenorean** [refraction-networking/utls#103]
I see you also developing in rust, so any plans to reimplement same functionality(on like a lite version) in rust(probably using boringssl or openssl)?

**Comment by gaukas** [comment]
Actually I liked and highly support this idea. 

If it is challenging to build a Rust library from rustls that supports ALL features in uTLS, at least it would be valuable to have a library to allow customization to a certain extend, or being able to (almost) perfectly mimic one, or a few, popular TLS implementations/browsers.

**Comment by gaukas** [comment]
Any early proposal or project draft is welcome and wanted!

**Comment by Gowee** [comment]
I patched rustls for that purpose for [noisy-shuttle](https://github.com/Gowee/noisy-shuttle). To keep to the changes minimum, I add a new [`rustls::ClientConnection::new_with`](https://github.com/Gowee/rustls-patched/blob/b15379697711631e5d46c893b65467bd1fa60a12/rustls/src/client/client_conn.rs#L566) that takes a custom client random, a custom session id and a function as an additional param which takes a `rustls::Message` and [modifies it](https://github.com/Gowee/noisy-shuttle/blob/fb844b9a7fc00a0d43388601eeaa5bee5392af44/tunnel/src/fp.rs#L144) per fingerprints. The implementation isn't porting features from utls while solely exists as a simplified alternative for my needs. And now it is a little outdated.

Porting utls to Rust would be a great idea. I am willing to contribute if anyone is gonna start such. Or I could start a project myself, but I may not be able to maintain it in a timely fashion on my own.

**Comment by gaukas** [comment]
Thanks for sharing @Gowee! Your fork looks great and indeed contains many features that a `rutls` library could benefit from.  

> I am willing to contribute if anyone is gonna start such. 

Sure, I don't mind kick this off by forking a popular and mostly-complete TLS library in Rust. Is Rustls a good starting point or what are the other choices? I'm not super familiar with the TLS ecosystem in Rust. 

> I may not be able to maintain it in a timely fashion on my own.

That is totally fine. Either way I don't believe on the refraction side we have enough capacity for `rutls`, so it will be more likely a community driven project. We could start it here under refraction's github org for now and decide its fate later.

**Comment by RPRX** [comment]
对 Rust 来说，或许直接使用 Chromium 的代码？https://github.com/XTLS/Xray-core/issues/2257#issuecomment-1606148760

**Comment by fedosgad** [comment]
AFAIK, @mhils (author of [mitmproxy](https://github.com/mitmproxy/mitmproxy/)) has looked for Rust analog of utls around time of mitmproxy v8 release. Maybe he knows something useful about such libraries?

**Comment by gaukas** [comment]
Thanks for tagging/linking @fedosgad. Out of curiosity, (how) did mitmproxy solve the TLS ClientHello fingerprinting issue in the end?

**Comment by mhils** [comment]
We don't have ClientHello mimicry in mitmproxy yet. Right now we're using OpenSSL via pyca/cryptography, which makes this pretty hard to implement. I looked into switching to rustls a while ago, but that attempt was abandoned due to compatibility issues (unrelated to mimicry) back then.

**Comment by gaukas** [comment]
> we're using OpenSSL via pyca/cryptography

Does it indicate that mitmproxy produces a relatively unique/unpopular ClientHello fingerprint or not? (Better if would like to provide it in [tlsfingerprint.io](https://tlsfingerprint.io/)'s format, you may pull from `https://client.tlsfingerprint.io/`, so we could examine how relatively popular it is)

> compatibility issues

I would entertain your insights on this if you would love to elaborate. What kind of compatibility issue and if there are any other good paths Rust (and potentially other programming language) programmer should take to counter the ClientHello fingerprinting?

i.e., if not then perhaps we should try to start building a rustls-based `rutls`, or alternatively, as @RPRX and @raintean mentioned here, a chromium/cronet Rust binding. 

(Disclaimer: refraction is not against such project, but unlikely to spare any capacity on leading/moderating the project. I would suggest starting a new organization and adding interested developers in, unless there is already something we could contribute to.)

**Comment by mhils** [comment]
> Does it indicate that mitmproxy produces a relatively unique/unpopular ClientHello fingerprint or not? (Better if would like to provide it in [tlsfingerprint.io](https://tlsfingerprint.io/)'s format, you may pull from `https://client.tlsfingerprint.io/`, so we could examine how relatively popular it is)

```shell
$ mitmdump --mode reverse:https://client.tlsfingerprint.io/ &
$ curl -k http://localhost:8080
{"tls_record_version":769,"tls_handshake_version":771,"cipher_suites":[4866,4867,4865,49195,49199,49196,49200,52393,52392,158,159,52394,49187,49191,49161,49171,49188,49192,49162,49172,103,107,156,157,60,61,47,53,255],"compression_methods":[0],"extensions":[0,11,10,35,22,23,13,43,45,51],"extensions_normalized":[0,10,11,13,22,23,35,43,45,51],"server_name":"client.tlsfingerprint.io","supported_groups":[29,23,30,25,24,256,257,258,259,260],"ec_point_formats":[0,1,2],"signature_algorithms":[1027,1283,1539,2055,2056,2057,2058,2059,2052,2053,2054,1025,1281,1537,771,769,770,1026,1282,1538],"alpn":null,"compress_certificate":null,"record_size_limit":[],"supported_versions":[772,771],"psk_key_exchange_modes":[1],"key_share":[29],"application_settings":null,"user_agent":"curl/7.81.0","nid":5075520528980946569,"norm_nid":-7307055319671900602,"id":"466fdf07a41cae89","norm_id":"9a981fe6b4d3f646"}
```

> I would entertain your insights on this if you would love to elaborate. What kind of compatibility issue and if there are any other good paths Rust (and potentially other programming language) programmer should take to counter the ClientHello fingerprinting?

One major blocker for us back then was the lack of support for IPs as dnsAltNames. Rustls supports that nowadays though. :)
Another issue for us was the lack of support for older TLS versions. We have users who put mitmproxy in front of outdated IoT devices to upgrade them to a recent TLS version. Typically not a concern otherwise.

**Comment by gaukas** [comment]
https://tlsfingerprint.io/id/N/9a981fe6b4d3f646

Unfortunately it does look like relatively unique and niche, with no known User Agents to us at our database... Our observation could be biased since it is from a university.

**Comment by mhils** [comment]
Thanks. I can also offer
```
{"tls_record_version":769,"tls_handshake_version":771,"cipher_suites":[4866,4867,4865,49195,49199,49196,49200,52393,52392,158,159,52394,49187,49191,49161,49171,49188,49192,49162,49172,103,107,156,157,60,61,47,53,255],"compression_methods":[0],"extensions":[0,11,10,35,16,22,23,13,43,45,51],"extensions_normalized":[0,10,11,13,16,22,23,35,43,45,51],"server_name":"client.tlsfingerprint.io","supported_groups":[29,23,30,25,24,256,257,258,259,260],"ec_point_formats":[0,1,2],"signature_algorithms":[1027,1283,1539,2055,2056,2057,2058,2059,2052,2053,2054,1025,1281,1537,771,769,770,1026,1282,1538],"alpn":["h2","http/1.1"],"compress_certificate":null,"record_size_limit":[],"supported_versions":[772,771],"psk_key_exchange_modes":[1],"key_share":[29],"application_settings":null,"user_agent":"curl/7.81.0","nid":-8632760564826273281,"norm_nid":1256227648061341672,"id":"883246012e8789ff","norm_id":"116f0463dfb87fe8"}
```
which you get if mitmproxy sends an ALPN because the client sent one. 😅

But yeah - I expect mitmproxy to be easily fingerprintable right now. I'd be happy to fix that eventually, but I'm prioritizing paid work at the moment. :)

**Comment by RPRX** [comment]
从工作量的角度来说，直接绑定 Chromium 的网络栈前期可能麻烦些，但后期的维护成本比较低，且可以获得比较好的效果。

修改 rustls，前期比较简单，但后期的维护成本比较高，会遇到一些 utls 遇到过的问题：
- 若 rustls 不支持 BoringSSL 的某些新功能，我们得自己实现
- 一些行为也要参考 Chrome 去进行修改
- 即使这样，还是会有很多细节无法做到一致

所以我个人倾向于直接绑定 Chromium 的网络栈，且由于它的维护成本低，放在该 org 下应该也不会有太大的负担。

**Comment by fedosgad** [comment]
> But yeah - I expect mitmproxy to be easily fingerprintable right now. I'd be happy to fix that eventually, but I'm prioritizing paid work at the moment. :)

I can confirm that mitmproxy is easily fingerprintable - there are DDoS protection services (Cloudflare, for example) that use ClientHello fingerprints to do their job. I have encountered websites that do not allow user to access them when using mitmproxy but work fine without it. That was the primary reason why I wrote [mirror_proxy](https://github.com/fedosgad/mirror_proxy/) a while ago.

**Comment by RPRX** [comment]
> Actually try to make a rust binding for cronet and you'll be fine. It's not hard for rust. But one thing that occurred to me while I was doing it is that it is very difficult to provide multiple versions of cronet in the same program. In other words, the library has to be compatible with one version of Chrome. That's not very scientific.

这应该不成问题，通常我们只用最新的 Chrome 指纹，你的库有计划公开吗，大概什么时候可以发布

**Comment by raintean** [comment]
> 这应该不成问题，通常我们只用最新的 Chrome 指纹，你的库有计划公开吗，大概什么时候可以发布？@raintean

@RPRX 目前只是个人实验项目, 暂时没有公开的计划.

**Comment by Gowee** [comment]
Some relevant discussions are here: https://github.com/rustls/rustls/issues/1125.

**Comment by gaukas** [comment]
Not sure why those infrastructure-level libraries are usually so reluctant or late to accept privacy/anti-censorship related features. 

But anyways, I think the needs for a mimicry library is well established. 

The next question is, in which form do we want it (single parrot or customizable parrots, i.e., Chrome or uTLS), and who's gonna volunteer to help out on and maintain the project.

**Comment by noorjs** [comment]
Building a wrapper for cronet only makes sense for the short term since it will be easier to maintain, there is already some open sourced code to help get started. However, building an independent TLS lib, i.e. forking boringSSL, rustls, etc makes the most sense for the long term. Most browsers are using some variation of openSSL, so maintaining a fork of openSSL (or boringSSL) with features found in uTLS seems to be the end game.

**Comment by gaukas** [comment]
A possibly off-topic update to this issue: at Refraction Networking we have recently created [water](https://github.com/gaukas/water) (it is under me only because it is yet to be cleaned up and formally adopted by Refraction Networking) and it has a twin brother in Rust [water-rs](https://github.com/erikziyunchi/water-rs). 

If everything goes well, we will have the ability to build uTLS into a WATER-compatible, pluggable WebAssembly Transport Module (WATM) and use it in Rust with `water-rs`. 

For now, we are blocking on TinyGo and Go. 

- TinyGo: lacks ability to correctly compile `crypto/tls`: https://tinygo.org/docs/reference/lang-support/stdlib/#cryptotls, which has only less dependencies than uTLS itself.
- Go: lacks ability to export WebAssembly functions (pending proposals available on `//go:wasmexport` directive)

If either one of these two could be unblocked at some point, we can advance our progress on building `uTLS.wasm` as an alternative solution to _uTLS in Rust_.

**Comment by gaukas** [comment]
> Update: We found a workaround and are able to build uTLS to a WebAssembly Transport Module which can be used by [water](https://github.com/gaukas/water). (12MB)

Another update: We were able to further reduce the size down to ~3MB. See [watm](https://github.com/gaukas/watm)'s release.

**Comment by kyochikuto** [comment]
Good work on [watm](https://github.com/gaukas/watm), but haven't projects such as [restls](https://github.com/3andne/restls) and [noisy-shuttle](https://github.com/Gowee/noisy-shuttle) already made this possible in Rust?

**Comment by gaukas** [comment]
> as restls and noisy-shuttle

I'm not familiar with noisy-shuttle, but I'm very sure restls implemented the client in Go to use uTLS's TLS ClientHello parrots. 

If you could point me to the TLS client implementation of noisy-shuttle, I would be very interested in taking a look.

**Comment by kyochikuto** [comment]
> > as restls and noisy-shuttle
> 
> I'm not familiar with noisy-shuttle, but I'm very sure restls implemented the client in Go to use uTLS's TLS ClientHello parrots.
> 
> If you could point me to the TLS client implementation of noisy-shuttle, I would be very interested in taking a look.

You're right restls does not depend on JA3 my mistake! But for noisy-shuttle, here's the fingerprint builder in [opts.rs](https://github.com/Gowee/noisy-shuttle/blob/fb844b9a7fc00a0d43388601eeaa5bee5392af44/shuttle/src/opt.rs#L113) and the implementation seems to be in the [fp.rs](https://github.com/Gowee/noisy-shuttle/blob/main/tunnel/src/fp.rs).

**Comment by gaukas** [comment]
Interesting! Thanks for the info, I will take a look into this, especially how they handle new/unimplemented extensions. 

But also it seems that `noisy-shuttle` itself is not doing a "real" TLS handshake, since some non-deprecated fields in the client hello was used for other purposes. Anyways I believe it should be at least partially meaningful towards a `utls-rust`.
